Overlay input is mixed-dimension for BDTOPO_V3 code insee 69123

Thank you for your reply. I added the line "srid":2154 to my configuration file. It worked with some "communes" but not with all ... For example, it worked with the commune "69200", but with others, such as the commune "69123" it didn't work, I got the following message:
 
Exception lors de l'appel de la fonction définie par l'utilisateur: "intersection(GEOMETRYCOLLECTION (POLYGON ((842137.7 
6515153.1, 842166.3999999997 6515153.1, 842165.6 6515152.3, 842167.1 6515150.7, 842163.9 6515148, 842165.2 6515146.6, 842162.5 6515144.2, 842164.8 6515141.6, 842166.9 6515139.3, 842168 6515138.1, 842159.4 6515130.3, 842158.3 6515131.5, 842156.2 6515133.9, 842156 6515133.8, 842151.8 6515138.5, 842150.9 6515137.6, 842147.1 6515134.3, 842137.7 6515125.857407407, 842137.7 6515153.1)), POINT (842137.7 6515123.1)), POLYGON ((842037.7 6515053.1, 842137.7 6515053.1, 842137.7 6515153.1, 842037.7 6515153.1, 842037.7 6515053.1))): Overlay input is mixed-dimension"; SQL statement: DROP TABLE IF EXISTS upper_table_join [90105-224]

How can I resolve this issue ? 

Thanks again in advance,

_Originally posted by @atoulou in https://github.com/orbisgis/geoclimate/discussions/931#discussioncomment-8367337_

Same issue with BD Topo V3.0 tested on 56185 and another municipality : 


```
The BDTOPO_V3 workflow has been started.
Please wait...
11:18:38.956 [main] INFO GeoClimate -- Processing the commune with the code insee :  56121
11:18:38.960 [main] INFO GeoClimate -- The location 56121 is represented by 2 polygons
. GeoClimate will process each polygon individually.
11:18:38.963 [main] INFO GeoClimate -- Processing the polygon 1 on 2 
11:18:40.021 [main] INFO GeoClimate -- BDTOPO V2 GIS layers formated
11:18:40.036 [main] INFO GeoClimate -- Create the units of analysis...
11:18:40.211 [main] INFO GeoClimate -- Start computing the geoindicators...
11:18:40.213 [main] INFO GeoClimate -- Start computing building indicators...
11:18:40.300 [main] INFO GeoClimate -- Start computing block indicators...
11:18:40.407 [main] INFO GeoClimate -- Start computing RSU indicators...
11:18:40.455 [main] INFO GeoClimate -- Processing urban typology surface fraction calculation
11:18:40.458 [main] INFO GeoClimate -- Processing LCZ surface fraction indicators calculation
11:18:40.480 [main] WARN GeoClimate --  The column pop doesn't exist or should be numeric
11:18:40.663 [main] INFO GeoClimate -- Geoindicators calculation time: 0.254 s
11:18:40.664 [main] INFO GeoClimate -- Start computing Typology indicators...
11:18:40.668 [main] INFO GeoClimate --  The LCZ classification will be performed 
11:18:40.716 [main] INFO GeoClimate --  The URBAN TYPOLOGY classification is performed 
11:18:44.683 [main] INFO GeoClimate -- All geoindicators have been computed
11:18:44.690 [main] INFO GeoClimate -- 56121 has been processed
11:18:44.690 [main] INFO GeoClimate -- Processing the polygon 2 on 2 
11:18:54.712 [main] INFO GeoClimate -- BDTOPO V2 GIS layers formated
11:18:54.713 [main] INFO GeoClimate -- Create the units of analysis...
11:19:15.971 [main] INFO GeoClimate -- Start computing the geoindicators...
11:19:15.971 [main] INFO GeoClimate -- Start computing building indicators...
11:19:36.035 [main] INFO GeoClimate -- Start computing block indicators...
11:19:39.157 [main] INFO GeoClimate -- Start computing RSU indicators...
11:19:50.236 [main] INFO GeoClimate -- Processing urban typology surface fraction calculation
11:19:50.246 [main] INFO GeoClimate -- Processing LCZ surface fraction indicators calculation
11:19:50.545 [main] WARN GeoClimate --  The column pop doesn't exist or should be numeric
11:20:18.326 [main] INFO GeoClimate -- Geoindicators calculation time: 39.169 s
11:20:18.327 [main] INFO GeoClimate -- Start computing Typology indicators...
11:20:18.327 [main] INFO GeoClimate --  The LCZ classification will be performed 
11:20:18.660 [main] INFO GeoClimate --  The URBAN TYPOLOGY classification is performed 
11:20:33.331 [main] INFO GeoClimate -- All geoindicators have been computed
11:20:33.335 [main] INFO GeoClimate -- 56121 has been processed
11:20:44.194 [main] WARN GeoClimate --  The column pop doesn't exist or should be numeric
org.h2.jdbc.JdbcSQLNonTransientException: Exception lors de l'appel de la fonction définie par l'utilisateur: "intersection(GEOMETRYCOLLECTION (POLYGON ((221395.4 6756189.69047619, 221400.4 6756188.5, 221398.7 6756181.9, 221395.4 6756182.5, 221395.4 6756189.69047619)), POINT (221395.4 6756196.1)), POLYGON ((221295.4 6756165.9, 221395.4 6756165.9, 221395.4 6756265.9, 221295.4 6756265.9, 221295.4 6756165.9))): Overlay input is mixed-dimension"
Exception calling user-defined function: "intersection(GEOMETRYCOLLECTION (POLYGON ((221395.4 6756189.69047619, 221400.4 6756188.5, 221398.7 6756181.9, 221395.4 6756182.5, 221395.4 6756189.69047619)), POINT (221395.4 6756196.1)), POLYGON ((221295.4 6756165.9, 221395.4 6756165.9, 221395.4 6756265.9, 221295.4 6756265.9, 221295.4 6756165.9))): Overlay input is mixed-dimension"; SQL statement:

                              DROP TABLE IF EXISTS upper_table_join [90105-224]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:566)
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:489)
	at org.h2.message.DbException.get(DbException.java:212)
	at org.h2.message.DbException.convertInvocation(DbException.java:436)
	at org.h2.schema.FunctionAlias$JavaMethod.execute(FunctionAlias.java:508)
	at org.h2.schema.FunctionAlias$JavaMethod.getValue(FunctionAlias.java:345)
	at org.h2.expression.function.JavaFunction.getValue(JavaFunction.java:40)
	at org.h2.schema.FunctionAlias$JavaMethod.execute(FunctionAlias.java:453)
	at org.h2.schema.FunctionAlias$JavaMethod.getValue(FunctionAlias.java:345)
	at org.h2.expression.function.JavaFunction.getValue(JavaFunction.java:40)
	at org.h2.expression.Alias.getValue(Alias.java:37)
	at org.h2.command.query.Select$LazyResultQueryFlat.fetchNextRow(Select.java:1851)
	at org.h2.result.LazyResult.hasNext(LazyResult.java:78)
	at org.h2.result.FetchedResult.next(FetchedResult.java:34)
	at org.h2.command.query.Select.queryFlat(Select.java:728)
	at org.h2.command.query.Select.queryWithoutCache(Select.java:833)
	at org.h2.command.query.Query.queryWithoutCacheLazyCheck(Query.java:197)
	at org.h2.command.query.Query.query(Query.java:520)
	at org.h2.command.dml.Insert.insertRows(Insert.java:197)
	at org.h2.command.dml.Insert.update(Insert.java:135)
	at org.h2.command.dml.DataChangeStatement.update(DataChangeStatement.java:74)
	at org.h2.command.ddl.CreateTable.insertAsData(CreateTable.java:199)
	at org.h2.command.ddl.CreateTable.update(CreateTable.java:157)
	at org.h2.command.CommandContainer.update(CommandContainer.java:169)
	at org.h2.command.CommandList.executeRemaining(CommandList.java:58)
	at org.h2.command.CommandList.update(CommandList.java:66)
	at org.h2.command.Command.executeUpdate(Command.java:256)
	at org.h2.jdbc.JdbcStatement.executeInternal(JdbcStatement.java:262)
	at org.h2.jdbc.JdbcStatement.execute(JdbcStatement.java:231)
	at org.h2gis.utilities.wrapper.StatementWrapper.execute(StatementWrapper.java:109)
	at groovy.sql.Sql.execute(Sql.java:2368)
	at org.orbisgis.data.jdbc.JdbcDataSource.execute(JdbcDataSource.java:474)
	at org.codehaus.groovy.vmplugin.v8.IndyInterface.fromCache(IndyInterface.java:321)
	at org.orbisgis.geoclimate.geoindicators.GenericIndicators.upperScaleAreaStatistics(GenericIndicators.groovy:976)
	at org.codehaus.groovy.vmplugin.v8.IndyInterface.fromCache(IndyInterface.java:321)
	at org.orbisgis.geoclimate.geoindicators.WorkflowGeoIndicators.rasterizeIndicators(WorkflowGeoIndicators.groovy:1841)
	at org.codehaus.groovy.vmplugin.v8.IndyInterface.fromCache(IndyInterface.java:321)
	at org.orbisgis.geoclimate.bdtopo.AbstractBDTopoWorkflow.computeGridIndicators(AbstractBDTopoWorkflow.groovy:844)
	at org.codehaus.groovy.vmplugin.v8.IndyInterface.fromCache(IndyInterface.java:321)
	at org.orbisgis.geoclimate.bdtopo.AbstractBDTopoWorkflow.bdtopo_processing(AbstractBDTopoWorkflow.groovy:766)
	at org.orbisgis.geoclimate.bdtopo.AbstractBDTopoWorkflow.bdtopo_processing(AbstractBDTopoWorkflow.groovy)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:343)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:328)
	at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:342)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1008)
	at org.codehaus.groovy.vmplugin.v8.IndyInterface.fromCache(IndyInterface.java:321)
	at org.orbisgis.geoclimate.bdtopo.AbstractBDTopoWorkflow$_execute_closure1.doCall(AbstractBDTopoWorkflow.groovy:237)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:343)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:328)
	at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:279)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1008)
	at groovy.lang.Closure.call(Closure.java:433)
	at groovy.lang.Closure.call(Closure.java:422)
	at org.codehaus.groovy.runtime.DefaultGroovyMethods.each(DefaultGroovyMethods.java:2389)
	at org.codehaus.groovy.runtime.DefaultGroovyMethods.each(DefaultGroovyMethods.java:2374)
	at org.codehaus.groovy.runtime.DefaultGroovyMethods.each(DefaultGroovyMethods.java:2427)
	at org.codehaus.groovy.runtime.dgm$209.doMethodInvoke(Unknown Source)
	at org.codehaus.groovy.vmplugin.v8.IndyInterface.fromCache(IndyInterface.java:321)
	at org.orbisgis.geoclimate.bdtopo.AbstractBDTopoWorkflow.execute(AbstractBDTopoWorkflow.groovy:231)
	at org.codehaus.groovy.vmplugin.v8.IndyInterface.fromCache(IndyInterface.java:321)
	at org.orbisgis.geoclimate.bdtopo.BDTopo.v3(BDTopo.groovy:167)
	at org.codehaus.groovy.vmplugin.v8.IndyInterface.fromCache(IndyInterface.java:321)
	at org.orbisgis.geoclimate.Geoclimate.call(Geoclimate.groovy:114)
	at org.orbisgis.geoclimate.Geoclimate.call(Geoclimate.groovy)
	at picocli.CommandLine.executeUserObject(CommandLine.java:1953)
	at picocli.CommandLine.access$1300(CommandLine.java:145)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2358)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2352)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2314)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2179)
	at picocli.CommandLine$RunLast.execute(CommandLine.java:2316)
	at picocli.CommandLine.execute(CommandLine.java:2078)
	at org.codehaus.groovy.vmplugin.v8.IndyInterface.fromCache(IndyInterface.java:321)
	at org.orbisgis.geoclimate.Geoclimate.main(Geoclimate.groovy:134)
Caused by: java.lang.IllegalArgumentException: Overlay input is mixed-dimension
	at org.locationtech.jts.operation.overlayng.EdgeNodingBuilder.addGeometryCollection(EdgeNodingBuilder.java:247)
	at org.locationtech.jts.operation.overlayng.EdgeNodingBuilder.add(EdgeNodingBuilder.java:229)
	at org.locationtech.jts.operation.overlayng.EdgeNodingBuilder.build(EdgeNodingBuilder.java:163)
	at org.locationtech.jts.operation.overlayng.OverlayNG.nodeEdges(OverlayNG.java:541)
	at org.locationtech.jts.operation.overlayng.OverlayNG.computeEdgeOverlay(OverlayNG.java:495)
	at org.locationtech.jts.operation.overlayng.OverlayNG.getResult(OverlayNG.java:483)
	at org.locationtech.jts.operation.overlayng.OverlayNG.overlay(OverlayNG.java:248)
	at org.locationtech.jts.operation.overlayng.OverlayNGRobust.overlaySnapTol(OverlayNGRobust.java:274)
	at org.locationtech.jts.operation.overlayng.OverlayNGRobust.overlaySnapping(OverlayNGRobust.java:214)
	at org.locationtech.jts.operation.overlayng.OverlayNGRobust.overlaySnapTries(OverlayNGRobust.java:186)
	at org.locationtech.jts.operation.overlayng.OverlayNGRobust.overlay(OverlayNGRobust.java:153)
	at org.h2gis.functions.spatial.operators.ST_Intersection.intersection(ST_Intersection.java:64)
	at jdk.internal.reflect.GeneratedMethodAccessor23.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.h2.schema.FunctionAlias$JavaMethod.execute(FunctionAlias.java:495)
	... 77 more

```

--- a/bdtopo/src/main/groovy/org/orbisgis/geoclimate/bdtopo/AbstractBDTopoWorkflow.groovy
+++ b/bdtopo/src/main/groovy/org/orbisgis/geoclimate/bdtopo/AbstractBDTopoWorkflow.groovy
@@ -915,7 +915,7 @@ abstract class AbstractBDTopoWorkflow extends BDTopoUtils {
             def urban_areas = dataFormated.urban_areas
 
 
-            info "BDTOPO V2 GIS layers formated"
+            info "BDTOPO GIS layers formated"
 
             def rsu_indicators_params = processing_parameters.rsu_indicators
             def worldpop_indicators = processing_parameters.worldpop_indicators

--- a/bdtopo/src/test/groovy/org/orbisgis/geoclimate/bdtopo/WorkflowDebugTest.groovy
+++ b/bdtopo/src/test/groovy/org/orbisgis/geoclimate/bdtopo/WorkflowDebugTest.groovy
@@ -137,7 +137,7 @@ class WorkflowDebugTest {
     @Test
     void testIntegrationFolderInput() {
         def input_data = "/media/ebocher/Extreme SSD/bdtopo/bdtopo2/BDTOPO_2-2_TOUSTHEMES_SHP_LAMB93_D035_2018-09-25/BDTOPO/1_DONNEES_LIVRAISON_2018-11-00144/BDT_2-2_SHP_LAMB93_D035-ED182"
-        def locations = ["Gimont"]
+        def locations = [ "Redon"]
         String directory = "/tmp/bdtopo2"
         File dirFile = new File(directory)
         dirFile.delete()
@@ -158,20 +158,27 @@ class WorkflowDebugTest {
                 "parameters"  :
                         ["distance"       : 0,
                          rsu_indicators   : [
-                                 "indicatorUse": ["LCZ"]//, "UTRF"]
+                                 "indicatorUse": ["LCZ",  "UTRF",  "URBAN_TYPOLOGY"]
                          ],
-                         /*"grid_indicators": [
+                         "grid_indicators": [
                                  "x_size"    : 1000,
                                  "y_size"    : 1000,
-                                 "indicators": ["LCZ_FRACTION"]
-                         ]*/
+                                 "indicators" :["BUILDING_FRACTION", "BUILDING_HEIGHT", "BUILDING_POP",
+                                 "BUILDING_TYPE_FRACTION", "WATER_FRACTION", "VEGETATION_FRACTION",
+                                 "ROAD_FRACTION", "IMPERVIOUS_FRACTION", "FREE_EXTERNAL_FACADE_DENSITY",
+                                 "BUILDING_HEIGHT_WEIGHTED", "BUILDING_SURFACE_DENSITY",
+                                 "SEA_LAND_FRACTION", "ASPECT_RATIO", "SVF",
+                                 "HEIGHT_OF_ROUGHNESS_ELEMENTS", "TERRAIN_ROUGHNESS_CLASS",
+                                 "UTRF_AREA_FRACTION", "UTRF_FLOOR_AREA_FRACTION",
+                                 "LCZ_PRIMARY"]
+                         ]
                         ]
         ]
         //BDTopo.v2(bdTopoParameters)
 
-        input_data = "/home/ebocher/Téléchargements/BDTOPO_3-3_TOUSTHEMES_SHP_LAMB93_D032_2023-09-15/BDTOPO"
+        input_data = "/media/ebocher/Extreme SSD/bdtopo/bdtopo3/BDTOPO_3-0_TOUSTHEMES_SHP_LAMB93_D035_2022-09-15/BDTOPO"
 
-        directory = "/tmp/bdtopo3/result"
+        directory = "/tmp/bdtopo3"
         dirFile = new File(directory)
         dirFile.delete()
         dirFile.mkdir()

--- a/geoindicators/src/main/groovy/org/orbisgis/geoclimate/geoindicators/GenericIndicators.groovy
+++ b/geoindicators/src/main/groovy/org/orbisgis/geoclimate/geoindicators/GenericIndicators.groovy
@@ -967,8 +967,8 @@ String upperScaleAreaStatistics(JdbcDataSource datasource, String upperTableName
                               DROP TABLE IF EXISTS $spatialJoinTable;
                               CREATE TABLE $spatialJoinTable 
                               AS SELECT b.$upperColumnId, a.$lowerColumnName,
-                                        ST_AREA(ST_INTERSECTION(st_force2d(st_makevalid(a.$lowerGeometryColumn)), 
-                                        st_force2d(st_makevalid(b.$upperGeometryColumn)))) AS area
+                                        ST_AREA(ST_INTERSECTION(a.$lowerGeometryColumn, 
+                                        b.$upperGeometryColumn)) AS area
                               FROM $lowerTableName a, $upperTableName b
                               WHERE a.$lowerGeometryColumn && b.$upperGeometryColumn AND 
                               ST_INTERSECTS(st_force2d(a.$lowerGeometryColumn), st_force2d(b.$upperGeometryColumn));

--- a/geoindicators/src/main/groovy/org/orbisgis/geoclimate/geoindicators/WorkflowGeoIndicators.groovy
+++ b/geoindicators/src/main/groovy/org/orbisgis/geoclimate/geoindicators/WorkflowGeoIndicators.groovy
@@ -2098,7 +2098,7 @@ String cutBuilding(JdbcDataSource datasource, String grid, String building) {
         DROP TABLE IF EXISTS $buildingCutted;
         CREATE TABLE $buildingCutted as 
         SELECT *, ST_AREA(THE_GEOM) AS area from 
-        (SELECT a.* EXCEPT(the_geom), st_intersection(a.the_geom, b.the_geom) as the_geom, b.* EXCEPT(the_geom),
+        (SELECT a.* EXCEPT(the_geom), ST_CollectionExtract(st_intersection(a.the_geom, b.the_geom), 3) as the_geom, b.* EXCEPT(the_geom),
         (b.HEIGHT_WALL + b.HEIGHT_ROOF) / 2 AS BUILD_HEIGHT
         FROM $grid as a, $building as b where a.the_geom && b.the_geom and st_intersects(a.the_geom, b.the_geom)) 
         as foo

